42 fit gp model on existing population from optimization problem

flo-schu opened the pull request:
Implements a Surrogate class for fitting a GP on existing data from optimizations

### implements 
- [x] fitting the GP on an existing population of samples
- [x] estimating the objectives functions
- [x] estimating the feasible objectives space
- [x] conditioning the `optimization_problem` on fixed variables
- ~~find `x0` at the edges of the problem~~

### plots
- [x] conditional
- [x] feasible space
- [x] minimum boundary
- [x] partial dependence 
- ~~overlay contour plot with true population~~

### tests
- [x] test if surrogate model breaks or not
- [x] test convergence of surrogate model to known solutions

### implementation of surrogates for different optimiztation tasks
| :zap: priority |
|----------------|
- [x] linear inequality constraints
- [ ] linear equality constraints
- [x] (pseudo) multi-objective, 
- [x] non-linear constraints


### performance
- [x] implement algorithm that finds a very good starting point for the task
- ~~find a faster method to test if the conditioned problem contains feasible space (creating a hopsy problem needs a lot of time 1/3 of total runtime)~~ This depends currently on hopsy

### code quality
- [x] Docstrings (GPT add docstrings, include warnings and errors, use numpy style, imperative style)
- [ ] refactor `condition_optimization_problem` method to a separate class that is an `OptimizationProblem`

## open questions

- actually sampled points overlaid over mean + uncertainty of GP
  GP.predict(std=True) -> mu, sigma
- contours? Lösen über fill between mit verschiedenen quantilen

## partial dependence

the function `partial_dependence` of `sklearn` returns a value that is not the actual response of the objective function but another value. This behavior is not ideal if it should be overlaid over the true objective space and the true evaluated points.

## minimum boundary plot

This plot draws the minimum boundary of the objective function over a given parameter.
There are two main approaches to achieve this

1. drawn a large number of samples and iterate over a moving window that always returns the minimum point. This is a pragmatic approach but it will be inexact
2. create a linear space over the target parameter x that should be drawn. Then optimize wrt all other parameters and find minimum. Ideally a good starting point would be that point which is close to the ones that are already close to the minimum by weighting these can be points from the surrogate or the true model evaluation (but surrogate is probably smarter, because there will be more points available)

this problem is solved by conditioning the optimization problem on one variable. This is done by fixing the conditioning variable at a specific value and substituting that value in the matrix and moving it to the RHS (`b`). 

## closes additional issues

- closes #33

flo-schu committed:
implment Surrogate class for fitting a GP on existing data from optimizations

implements methods for:
- fitting the GP
- estimating the objectives functions
- estimating the feasible objectives space
- conditioning the GP

flo-schu commented:
## trying out new features

currently the fixtures for optimization problems are not on the `dev` branch. @schmoelder, maybe you can change this? 

Until then, there is a "fix". I am playing around with the surrogate model in the paper repository: https://github.com/flo-schu/bayesian-optimization
There I have just copy pasted the fixtures and work on the plots. If you clone this repo, you can try out the surrogate features

schmoelder commented:
> currently the fixtures for optimization problems are not on the `dev` branch. @schmoelder, maybe you can change this?

Yes, I will try to move it to `dev`.

flo-schu commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,119 @@
+            else:
+                free_vars.update({var.name: var_index})
+
+        X, F = self.approximate_objectives(n_samples=n_samples)
```
yes. I'll change it.

flo-schu commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,119 @@
+
+
+
+    def estimate_objectives(self, X):
```
I agree. But I wasn't sure about the design of this anyway. Maybe it would be better to design the functions/properties of Meta scores, constraint violations, ... analogous to what you did in the optimization problem / results? And only replace `evaluate` with `estimate`?

schmoelder replied:
Probably we should just mimic the `OptimizationProblem` interface for now. I.e. have individual functions for objectives, constraints, etc. This makes it easier to later fit it into existing architecture. 

Since at some point, I'd like to unify the `evaluate_` and evaluate_population` methods in the `OptimizationProblem` interface, I don't think we need to implement both here and already make the `evaluate` function work for both.

flo-schu committed:
fix method call `estimate_objectives` in `condition_objectives`.

schmoelder commented on CADETProcess/optimization/optimizationProblem.py:
```diff
@@ -143,6 +143,10 @@ def wrapper(self, x, *args, get_dependent_values=False, **kwargs):
 
         return wrapper
 
+    def get_variable_index(self, variable):
```
please add a docstring.

flo-schu replied:
done

schmoelder commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,119 @@
+        return X, F
+
+
+    def condition_objectives(
```
Also add function to condition constraints etc.

schmoelder commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,119 @@
+            var.lb = lb
+            var.ub = ub
+
+    def fit_gaussian_process(self, population: Population):
```
I'm wondering, maybe we should already pass a population in the constructor and provide an `update_model` function? This way we never encounter an un-initialized surrogate... But I guess it depends on where we put it in the architecture...

flo-schu replied:
already changed

flo-schu committed:
WIP: made a first approach to integrate Surrogate into post_processing and optimization results.

flo-schu commented:
I've made a very first approach for integration of Surrogate in `post_processing` and `OptimizationResults` it works for `U_NSGA` but it is still very WIPy.

Let's talk it through tomorrow and see which additional features we want and what can be improved, refactored, etc. :)

flo-schu committed:
modifications of implementation and TODOs

Co-authored-by: Johannes Schmölder <schmoelder@users.noreply.github.com>

flo-schu committed:
don't add population when initializing surrogate

flo-schu committed:
WIP: find mimum objective for a given x

flo-schu commented:
not sure if the merge of dev into 42... was the right thing to do. Probably rebase would have been your preferred choice? If possible you can walk me through correcting this and doing it differently.

flo-schu committed:
move create_hopsy_problem to separate method

flo-schu committed:
use hopsy problem to check feasibility (faster)

flo-schu committed:
update todos

flo-schu committed:
test problem for higher oder inequality linear constraints

flo-schu committed:
add comment for fixing a potential bug

flo-schu committed:
add docstrings

flo-schu committed:
remove surrogate from optimization results

flo-schu committed:
remove bug hint

not necessary, because sequentially removing and adding constraints at the same time is not a good idea at any rate

flo-schu committed:
clean up

flo-schu commented:
okay, I have cleaned up `surrogate` module and removed it from `OptimizationResults` and `Optimizer`. Next week I will continue developing it for further optimization tasks and write some tests

schmoelder commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,381 @@
+        M_est = self.surrogate_model_M.predict(X)
+        return M_est
+
+    def get_conditional_and_free_indices(self, conditional_vars={}):
```
Using mutable objects as default arguments is dangerous.

Better use 

```
def get_conditional_and_free_indices(self, conditional_vars=None):
    if conditional_vars = None:
        conditional_vars  = {}
```

Otherwise, all function calls will potentially refer to the same dict object.

schmoelder commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,381 @@
+
+    def __init__(
+        self,
+        optimization_problem: OptimizationProblem,
```
Only use OptimizationResults as parameter. The OptimizationProblem is also stored in that. This way, we can avoid inconsistencies.

schmoelder commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,381 @@
+
+        return cond_var_idx, free_var_idx
+
+    def condition_constraints(self, conditional_vars={}):
```
see above...

flo-schu replied:
done

schmoelder commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,381 @@
+        Initialize the Surrogate class.
+
+        Parameters:
+        - optimization_problem (OptimizationProblem): The optimization problem.
```
Please use numpy style docstrings.

E.g.

```
def __init__(self, optimization_problem, population, n_samples):
    """
    Initialize the Surrogate class.

    Parameters
    ----------
    optimization_problem : OptimizationProblem
        The optimization problem.
    """
    [...]
```

flo-schu commented on CADETProcess/optimization/surrogate.py:
```diff
@@ -0,0 +1,381 @@
+        return A_cond, b_cond
+
+
+    def condition_optimization_problem(self, conditional_vars={}):
```
done

flo-schu committed:
make change according to review

flo-schu committed:
implement finding of good starting value by interpolating f_minimal neighbors and refactor pairplot

flo-schu committed:
implement fill_gaps method and refactor

flo-schu committed:
Use euclidian distance on transformed space and refactor n_neighbors

allow only 1 or more neighbors on either side of the point of interest

Co-authored-by: Johannes Schmölder <schmoelder@users.noreply.github.com>

flo-schu committed:
implement  nonlinear constraints

flo-schu committed:
WIP: implement pseudo multi objective

flo-schu committed:
implement first surrogate tests

flo-schu commented:
wrapper around evaluate_objectives does not work because the optimization
problem can't deal with the new number of variabels

this has to do with `set_variables` and `get_dependent_variables` which return a lower number of variables than those that are actually present.

Therefore it is probably best to revert and use the old function (for now)
and find a way to constrain evaluation objects

Although this is messy it will probably easier yield a result than refactoring
condition optimization problem into something leaner and more proper 
whatever that means

flo-schu committed:
working SOO, MOO and NLC,

next stop evaluators

flo-schu committed:
refactor function conditioning, to harmonize it for nonlinear constraints and objectives

flo-schu committed:
always return multi-objective results of surrogate (ndim=2)

flo-schu committed:
improve tests

flo-schu committed:
refactor for returning individuals from surrogate if X is 1D to match CADETProcess

flo-schu committed:
implement conditioning with `Evaluator` objects.

- Works for single objective linear constraints
- Does not work for multi objective nonlinear constraints

flo-schu committed:
fix evaluator conditioning and log infeasibility errors

flo-schu committed:
new test for nonlinear and linear constraints moo.

Surprisingly it works

flo-schu committed:
hotfix for complicated problem in surrogate

flo-schu committed:
improve error messages

flo-schu committed:
undo hotfix because it broke other tests

flo-schu committed:
add evaluator to constraints to reproduce error in jos problem

flo-schu committed:
add evaluator to nonlinear_constraint test case

flo-schu committed:
implement evaluator conditioning for nonlinear constraints

flo-schu committed:
fix surrogate nlc indexing bug, implement estimate x0 method

flo-schu committed:
refactor: evaluators in surrogate only pass on and complete X

flo-schu committed:
implement central evaluator conditioning

flo-schu commented:
| :warning: GP instability for nonlinear constraints |
|----------------------------------------------------|

flo-schu committed:
implement estimation of constraint_violations

flo-schu committed:
new optimization_problem test for nonlinear soo

flo-schu committed:
WIP add conditioned nonlinear constraints via api

flo-schu commented:
While the behavior could be fixed, still the GP cannot be fitted on the problem. As suspected in:
> 2. standard scaling before fitting the GP. If the constraints/parameters operate on different scales, this might reduce the problem. This is unclear, if it will work, but it might improve the stability of the GP in any case. Then however, all X would need to be scaled before prediction (with a scaler provided to the surrogate, not to be confused with the transform method.

## other potential solutions:

- [ ] Scaling might also aid the problem, as F and G operate on highly different scales. Then standard scaling should be performed.
- [ ] Conceptually think about, whether it makes sense that nonlinear constraints can be violated in the same range where other individuals do not violate constraints.
- [ ] plot parameters / constraints predictions against true points

flo-schu committed:
fix sesimtate x0

surrogate for jos problem still does not work.

tests are fine

flo-schu committed:
implement standard scaling

flo-schu committed:
formatting

flo-schu committed:
formatting

flo-schu committed:
fix dimensionality problem for scaling, implement convergence test

flo-schu committed:
implement convergence testing of minimum boundary

flo-schu committed:
tests passing, convergence checks ok, bug supposedly fixed

flo-schu committed:
clean up test_surrogate

flo-schu committed:
refactor gp training

flo-schu committed:
refactor estimators by using a wrapper

flo-schu committed:
rename method condition_linear_constraints

flo-schu committed:
fix bug

flo-schu committed:
clean up x0 finding routine

flo-schu committed:
remove commented code blocks

flo-schu committed:
rename test_tol to test_abs_tol

flo-schu committed:
use np.testing.assert_allclose

flo-schu committed:
implement else action if shapes of rounded and unrounded problems do not match

flo-schu commented:
> Scaling might also aid the problem, as F and G operate on highly different scales. Then standard scaling should be performed.

scaling solved the problem entirely

flo-schu commented:
branch `jos-problem` is now merged into this branch and tests are passing. Currently the surrogate model does not work (again) on the `no-waste` problem. This is because `estimate_x0` was removed from the module. I will integrate a reduced version in `find_x0.` With this, hopefully, the chebyshev computations and hopsy can be removed entirely from the `find_minimum` method.

- [ ] implement x0-finding algorithm at the edges of the problem in `find_x0`

When this is done, only the `condition_optimization_problem` needs to be refactored into a separate class, but this could also be done later

- [ ] refactor `condition_optimization_problem`

flo-schu commented:
x0 finding at the edges of the problem is increasingly difficult. I have decided to roll back on the state with a more costly iterative x0 estimation technique.

With some more time, it would definitely be possible to find a good solution for finding x0 at the boundaries of the problem. 

Ideally however the following algorithmus can be used:

1. analyze if the conditioned problem contains feasible space (e.g. with hopsy) (implemented but problematic due to hopsy issues)
2. if not continue with next conditioning slice
3. if feasible space exist, use existing evaluations to interpolate a new x0 (implemented)
4. if feasible space exists, but no points are there to interpolate, extrapolate into unexplored space. 

For 4. caveats exist:

- the surrogate in the unexplored space may return very unlikely predictions that are out of bounds for the true constraints or bounds of the problem
- in this case it may be better to not go into these regions by extrapolating and restrict the surrogate to the sampled range (which will most likely be close enough to the bounds)
- If the borders must be inside the problem, it may be a good idea to sample the corners of the problem in the first place. Because then also it will be much easier for the surrogate find good points near the boundary of a problem

flo-schu commented:
currently the module passes all tests and also is enabled to solve the no-waste problem. It could be merged. 

Open issues and improvements are tracked in https://github.com/fau-advanced-separations/CADET-Process/issues/52

schmoelder commented:
I think rebasing worked. However, there is now an issue with Python < 3.9:

E.g.
```
======================================================================
ERROR: test_surrogate (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_surrogate
Traceback (most recent call last):
  File "/usr/share/miniconda3/envs/test/lib/python3.9/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/share/miniconda3/envs/test/lib/python3.9/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/home/runner/work/CADET-Process/CADET-Process/tests/test_surrogate.py", line 6, in <module>
    from CADETProcess.optimization import (
  File "/home/runner/work/CADET-Process/CADET-Process/CADETProcess/__init__.py", line 34, in <module>
    from . import optimization
  File "/home/runner/work/CADET-Process/CADET-Process/CADETProcess/optimization/__init__.py", line 97, in <module>
    from .surrogate import Surrogate
  File "/home/runner/work/CADET-Process/CADET-Process/CADETProcess/optimization/surrogate.py", line 22, in <module>
    class Surrogate:
  File "/home/runner/work/CADET-Process/CADET-Process/CADETProcess/optimization/surrogate.py", line 162, in Surrogate
    def estimate_objectives(self, X):
TypeError: 'staticmethod' object is not callable

```
I'm not entirely sure why this suddenly comes up but it might be due to the recent changes in the Structure base class.

For reference: https://stackoverflow.com/questions/41921255/staticmethod-object-is-not-callable

This issue seems to e resolved in Python 3.10 but it's probably too early to "retire" 3.9. So we might have to look into this again.

schmoelder committed:
Only evaluate objects registered in objective/constraint functions

schmoelder committed:
Fix objective names

schmoelder committed:
Restructure OptimizationProblem metrics

schmoelder committed:
Add `results` in example callback.

schmoelder committed:
Add support for creating x0 with linear equality constraints

schmoelder committed:
Allow setting empty list for output_state for units without destination

schmoelder committed:
Check cstr volume only when flow rate is not None

schmoelder committed:
Improve ReferenceIO solution shape check

Commit:
Fix calculation of conversion rates for MSSMA model

Commit:
Fix behavior of comparisons to references with multiple components

schmoelder committed:
Add n_difference_metrics property

schmoelder committed:
Improve figure styling

schmoelder committed:
Update CI

- Fix caching
- Update Python versions
- Test on MacOS

Commit:
Add examples for SMA parameter determination using Yamamoto and inverse method

schmoelder committed:
Raise warning if x0 is smaller than popsize

Commit:
Untransform x when evaluating callbacks.

Commit:
Clarify that OptimizationProblem.transform expects independent space

Commit:
Remove increased burn-in when using equality constraints

Commit:
Assume x0 contains only independent variables

Commit:
Return dependent variables for x0 by default

Commit:
Temporarily disable test_fit_column_parameters to avoid breaking CI

schmoelder committed:
Fix jupytext formatting

schmoelder committed:
Move chebyshev center to own method

schmoelder committed:
Improve docstring

schmoelder committed:
Add check_individual method

schmoelder committed:
Move validation of x0 to OptimizerBase
